include/RAJA/policy/hip/kernel/For.hpp

```diff
@@ -686,35 +545,36 @@ struct HipStatementExecutor<
   using diff_t = segment_diff_type<ArgumentId, Data>;
 
 
-  static
-  inline RAJA_DEVICE void exec(Data &data, bool thread_active)
+  static inline RAJA_DEVICE
+  void exec(Data &data, bool thread_active)
   {
     // grid stride loop
     diff_t len = segment_length<ArgumentId>(data);
-    diff_t i_init = get_hip_dim<BlockDim>(dim3(blockIdx.x,blockIdx.y,blockIdx.z));
-    diff_t i_stride = get_hip_dim<BlockDim>(dim3(gridDim.x,gridDim.y,gridDim.z));
+    diff_t i_init = Indexer::template index<diff_t>();
+    diff_t i_stride = Indexer::template size<diff_t>();
 
-    // Iterate through grid stride of chunks
-    for(diff_t i = i_init;i < len;i += i_stride){
+    // Iterate through chunks
+    for (diff_t ii = 0; ii < len; ii += i_stride) {
+      diff_t i = ii + i_init;
 
-      // Assign the x thread to the argument
+      // execute enclosed statements if any thread will
+      // but mask off threads without work
+      bool have_work = i < len;
+
+      // Assign the index to the argument
       data.template assign_offset<ArgumentId>(i);
 
       // execute enclosed statements
-      enclosed_stmts_t::exec(data, thread_active);
+      enclosed_stmts_t::exec(data, thread_active && have_work);
```

Comment:
What is the difference here between thread_active and have_work? Is there way to combine them into a single mask? Might need to check if having both Having both causes additional conditional checks.

Reply:
There is a difference between them. What is called thread_active is used to ensure all threads in a block participate so collective operations like syncthreads are valid. What is called have_work is used to mask out threads at the statement::Lambda so only those that are in bounds/have work run.